docs/architecture-decisions/decouple-flag-source-and-set.md

```diff
@@ -0,0 +1,77 @@
+---
+status: proposed
+author: @tangenti
+created: 2025-06-16
+updated: 2025-06-16
+---
+
+# Decouple flag sync sources and flag sets
+
+The goal is to support dynamic flag sets for flagd providers and decouple sources and flag sets.
+
+## Background
+
+Flagd daemon syncs flag configurations from multiple sources. A single source provides a single config, which has an optional flag set ID that may or may not change in the following syncs of the same source.
```

Comment:
> which has an optional flag set ID that may or may not change in the following syncs of the same source.

This is a good point. I think we need to make sure we _do_ support changing this value and all the expected behavior associated with that.

Reply:
@tangenti for example, if the `flagSetId` is removed from a flag, we probably need to recompute the sets for listeners on that flagSet not to include that flag now. To put it another way, changes in the flag set could result in effective removal from listener sets. This isn't a problem, but we need to make sure we implement it this way, I think.